src/peft/utils/hotswap.py

```diff
@@ -13,23 +13,302 @@
+    elif isinstance(lora_module.scaling[adapter_name], (float, int)):
+        lora_module.scaling[adapter_name] = scaling
+    else:
+        raise ValueError(
+            "Something went wrong when trying to set the new scale value, expected to find the old value to be of type "
+            f"float or torch.Tensor, got {type(lora_module.scaling[adapter_name])} instead."
+        )
+
+
+def _convert_scalings_to_tensor(model):
+    """
+    Convert the LoRA scaling values into torch.tensors to prevent recompilation if they change.
+    """
+    for module in model.modules():
+        if not isinstance(module, LoraLayer):
+            continue
+
+        scaling = module.scaling
+        for key, val in scaling.items():
+            if isinstance(val, float):
+                scaling[key] = torch.tensor(val, device=module.weight.device)
+            elif not isinstance(val, torch.Tensor):
+                raise ValueError(
+                    "Something went wrong while trying to convert the scalings, expected to find values of type float "
+                    f"but found {type(val)} instead."
+                )
+
+
+def _get_padded_linear(lora_module: torch.nn.Module, target_rank: int, is_lora_A: bool) -> torch.nn.Linear:
+    """
+    Get a new Linear layer for LoRA with padded weights according to the target rank.
+
+    Args:
+        lora_module (nn.Module):
+            The LoRA sub-module (e.g. module.lora_A[adapter_name]).
+        target_rank (int):
+            The desired rank to pad to.
+        is_lora_A (bool):
+            True if this is the LoRA A matrix, False if LoRA B.
+
+    Returns:
+        nn.Linear:
+            A newly created and padded Linear layer. If the rank already fit, the original layer is returned.
+    """
+    weight = lora_module.weight
+    # For LoRA A, the "rank dimension" is weight.size(0) (out_features).
+    # For LoRA B, it is weight.size(1) (in_features).
+    original_rank = weight.size(0) if is_lora_A else weight.size(1)
+
+    # If no padding needed
+    if original_rank == target_rank:
+        return lora_module
+
+    if original_rank > target_rank:
+        raise ValueError(
+            f"Trying to pad the adapter to the target rank {target_rank}, but the original rank is larger "
+            f"({original_rank}). This is not possible."
+        )
+
+    out_features, in_features = weight.shape
+
+    if is_lora_A:
+        # LoRA A affects out_features
+        padded = torch.zeros(target_rank, in_features, device=weight.device, dtype=weight.dtype)
+        padded[:original_rank, :] = weight
+        new_layer = torch.nn.Linear(in_features, target_rank, bias=lora_module.bias is not None)
+    else:
+        # LoRA B affects in_features
+        padded = torch.zeros(out_features, target_rank, device=weight.device, dtype=weight.dtype)
+        padded[:, :original_rank] = weight
+        new_layer = torch.nn.Linear(target_rank, out_features, bias=lora_module.bias is not None)
```

Comment:
I would also add a comment to note that we might revisit if we have to consider if the corresponding layers come from a quantization backend.

Reply:
I am thinking about functions like this:
https://github.com/huggingface/peft/blob/2825774d2de1c8bd0604ac685867edd79d608a9e/src/peft/tuners/lora/bnb.py#L546